bench/macro/lsm-tree-bench-union.hs

```diff
@@ -0,0 +1,1001 @@
+
+        LSM.withIncrementalUnions tables $ \table -> do
+          LSM.UnionDebt totalDebt <- LSM.remainingUnionDebt table
+          -- Determine the number of credits to supply per tick
+          -- in order to have all debt repaid when 90% complete.
+          let paymentPerTick = ceiling $ product
+                [ paymentRate,  toInteger totalDebt % 1, 1 % steps ]
+          let tickCredit = LSM.UnionCredits paymentPerTick
+          forM [1 .. steps] $ \step -> do
+            void $ LSM.supplyUnionCredits table tickCredit
+            LSM.UnionDebt _debt <- LSM.remainingUnionDebt table
+            (time, _, _) <- timed_ $ do
+              benchmarkIterations
+                h
+                (\_ _ -> pure ())
+                (initialSize gopts)
+                (batchSize opts)
+                (batchCount opts)
+                (seed opts)
+                table
+            -- Perform measurement of batched lookups
+            -- Save the result for later to be included in the performance plot
+            let ops = batchCount opts * batchSize opts
+                rate = fromIntegral ops / time
+            -- Print a status report while running the benchmark
+            printf
+              (Fold.fold [
+                "    [%",
+                show . length $ show steps,
+                "d/",
+                show steps,
+                "]:    %7.01f ops/sec",
+                "    with debt = %8d\n"
+              ])
+              step
+              rate
+              _debt
+            pure rate
+
+    -- Generate a performance plot based on the benchmark results.
+    Plot.toFile Plot.def (rootDir gopts <> "/benchmark.png") $ do
+      Plot.layout_title .= "Incremental Unions Performance"
+      Plot.layout_x_axis . Plot.laxis_override .= Plot.axisGridHide
+      Plot.layout_x_axis . Plot.laxis_title    .= "Credits supplied over time"
+      Plot.layout_y_axis . Plot.laxis_title    .= "Lookup access time"
+      Plot.plot $ Plot.line "operations per second" [ zip [1 :: Word .. ] measurements ]
+
+-------------------------------------------------------------------------------
+-- sequential
+-------------------------------------------------------------------------------
+
+type LookupResults = V.Vector (K, Maybe V)
+
+{-# INLINE sequentialIteration #-}
+sequentialIteration :: LatencyHandle
+                    -> (Int -> LookupResults -> IO ())
+                    -> Int
+                    -> Int
+                    -> LSM.Table K V
+                    -> Int
+                    -> MCG.MCG
+                    -> IO MCG.MCG
+sequentialIteration h output !initialSize !batchSize !tbl !b !g =
+    withTimedBatch h b $ \tref -> do
+    let (!g', ls, is) = generateBatch initialSize batchSize g b
+
+    -- lookups
+    results <- timeLatency tref $ LSM.lookups tbl ls
+    output b (V.zip ls (fmap (fmap (const mempty)) results))
+
+    -- deletes and inserts
+    _ <- timeLatency tref $ LSM.updates tbl is
```

Comment:
@jorisdral I think that you are correct, insertion and/or updates are being performed in the benchmark. I'll have to remove this.

Reply:
All insertion/updates are removed.